### PR TITLE
VDS-437: Implement the backend for the configured product parts

### PR DESCRIPTION
**Pull request by trueboroda**

### CODE REVIEW

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Cart/DesignTimeCartDbContextFactory.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ public DemoCartDbContext CreateDbContext(string[] args)
         {
             var builder = new DbContextOptionsBuilder<DemoCartDbContext>();
 
-            builder.UseSqlServer("Data Source=(local);Initial Catalog=VirtoCommerce3;Persist Security Info=True;User ID=virto;Password=virto;MultipleActiveResultSets=True;Connect Timeout=30");
+            builder.UseSqlServer("Data Source=(local);Initial Cart=VirtoCommerce3;Persist Security Info=True;User ID=virto;Password=virto;MultipleActiveResultSets=True;Connect Timeout=30");
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Cart/DesignTimeCartDbContextFactory.cs]: Revert.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoProductPartEntity.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            ModifiedDate = part.ModifiedDate;
+            ModifiedBy = part.ModifiedBy;
+
+            ConfiguredProductId = part.ConfiguredProductId;
+            Name = part.Name;
+            ImgSrc = part.ImgSrc;
+            Priority = part.Priority;
+            DefaultItemId = part.DefaultItemId;
+
+            if (part.ItemsIds != null)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoProductPartEntity.cs:81]: Use IsNullOrEmpty extension

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoProductPartEntity.cs (hunk 2 above)
Comment (trueboroda) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoProductPartEntity.cs:81]: The null value is mean nothing to do.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPart.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+        public string Name { get; set; }
+
+        public string ImgSrc { get; set; }
+
+        public int Priority { get; set; }
+
+        public string DefaultItemId { get; set; }
+
+        public string[] ItemsIds { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPart.cs:18]: Initialize collection in the constructor

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            , IAuthorizationService authorizationService)
+        {
+            _partsSerarchService = partsSerarchService;
+            _partsService = partsService;
+            _authorizationService = authorizationService;
+        }
+
+        [HttpGet]
+        [Route("product/parts/{id}")]
+        public async Task<ActionResult<DemoProductPart>> Search(string id)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs]: Rename method to GetProductPartById

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            var result = (await _partsService.GetByIdsAsync(new[] { id })).FirstOrDefault();
+            return Ok(result);
+        }
+
+        [HttpPost]
+        [Route("product/parts/search")]
+        public async Task<ActionResult<DemoProductPartSearchResult>> Search(DemoProductPartSearchCriteria criteria)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs:38]: Add Authorize attribute to each action

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            var result = await _partsSerarchService.SearchProductPartsAsync(criteria);
+            return Ok(result);
+        }
+
+        [HttpPost]
+        [Route("product/parts")]
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs:45]: Add ProducesResponseType attribute to each action when it needed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Catalog/DemoProductPartService.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            Func<ICatalogRepository> catalogRepositoryFactory,
+            IPlatformMemoryCache platformMemoryCache,
+            IEventPublisher eventPublisher)
+        {
+            _repositoryFactory = catalogRepositoryFactory;
+            _platformMemoryCache = platformMemoryCache;
+            _eventPublisher = eventPublisher;
+        }
+
+        public async Task<DemoProductPart[]> GetByIdsAsync(string[] partIds)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Catalog/DemoProductPartService.cs]: Make all public methods virtual

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Review by asvishnyakov — DISMISSED** [review]
I have nothing to add. Agree with issues found by @pushnitsa 
Security hotspot should resolved manually

**Review by pushnitsa — DISMISSED** [review]

**Review by ilyawzrd — APPROVED** [review]

**Pull request merged**